### PR TITLE
QLP Helper Functionality

### PR DESCRIPTION
**Pull request by havijw**
These changes re-introduce helper functions for running bulk predictions in a context like QLP, including assembling datasets, loading model checkpoints from a directory, and running predictions using the models on the dataset.

A significant change is the introduction of `astronet.direct_tensor`, which allows users to create a dataset without having to write `TFRecord` files as a side effect. This is much more efficient for running one-off predictions in bulk.

### CODE REVIEW

File: astronet/direct_tensor/features.py
Comment (havijw) [inline astronet/direct_tensor/features.py:1]: Needs documentation

File: astronet/direct_tensor/predict.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+    det_time, det_flux, transit_mask = preprocess.detrend_and_filter(
+        tic, time, flux, period, epoch, duration, breakspace
+    )
+    folded_time, folded_flux, fold_num, normal_transit_mask = (
+        preprocess.phase_fold_and_sort_light_curve(
+            det_time, det_flux, transit_mask, period, epoch
+        )
+    )
+    odd_mask = fold_num % 2 == 1
+    even_mask = fold_num % 2 == 1
```
Comment (havijw) [inline astronet/direct_tensor/predict.py]: These should not be the same!

File: astronet/direct_tensor/features.py
Comment (havijw) [inline astronet/direct_tensor/features.py:1]: Needs documentation

File: astronet/direct_tensor/predict.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        "star_mass_present": float(np.isnan(tce["SMass"])),
+        "star_rad": tce["SRad"] if not np.isnan(tce["SRad"]) else 0.0,
+        "star_rad_present": float(np.isnan(tce["SRad"])),
+        "star_rad_est": tce["SRadEst"] if not np.isnan(tce["SRadEst"]) else 0.0,
+        "star_rad_est_present": float(np.isnan(tce["SRadEst"])),
+        "n_folds": len(set(fold_num)),
+        "n_points": len(fold_num),
+    }
+    for feature, value in scalar_features.items():
+        assert not np.isnan(
```
Comment (WilliamCFong) [inline astronet/direct_tensor/predict.py]: We'll probably want to raise something other than an assert error. If anyone runs `python -O`, this line will not raise.

File: astronet/direct_tensor/predict.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        tried and the best selected. Default = [0.3, 5.0, None].
+
+    Returns
+    -------
+    features: dict[str, float | array[float]]
+        Dict of {feature_name: feature_value} for all inputs for the lightcurve.
+    """
+    all_features = {}
+
+    for breakspace in breakspaces:
```
Comment (havijw) [inline astronet/direct_tensor/predict.py]: If `breakspaces` is empty, then `fold_num` will never be set, which is a problem for the scalar features. If breakspaces is non-empty, we should just run whatever function calculates `fold_num` in `standard_view_features`. If `fold_num` has different values for different breakspaces, we should raise a `RuntimeError`.

File: astronet/predict.py [hunk 4]
```diff
@@ -90,15 +93,86 @@ def predict(legacy=False):
...
-    if FLAGS.output_file:
-      with tf.io.gfile.GFile(FLAGS.output_file, "w") as f:
+    if output_file:
+      with tf.io.gfile.GFile(output_file, "w") as f:
         results.to_csv(f)
         
     return results, config
 
 
+def checkpoint_dirs(root: str, nruns: Optional[int] = None) -> list[str]:
```
Comment (WilliamCFong) [inline astronet/predict.py:104]: Could we use your new checkpoint function? This one is a bit archaic with the `os.path` usage and the runtime checks.

File: astronet/direct_tensor/predict.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    aperture_fluxes: dict[str, tuple[np.ndarray, np.ndarray]],
+    breakspaces: list[Optional[float]] = [0.3, 5.0, None],
+):
+    """
+    Assemble all input features for a lightcurve.
+
+    Params
+    ------
+    tce: pd.Series
+        Row of dataframe containing relevant TCE parameters. Should include "Astro ID", "Per",
```
Comment (havijw) [inline astronet/direct_tensor/predict.py]: Nit: docstrings should be capped at 80 character width, not 100

File: astronet/predict.py [hunk 6]
```diff
@@ -90,15 +93,86 @@ def predict(legacy=False):
...
+    for features, identifiers in ds:
+      preds = model(features)
+
+      row = {}
+      row['astro_id'] = identifiers.numpy().item()
+      for i, p in enumerate(preds.numpy()[0]):
+        row[label_index[i]] = p
+
+      series.append(row)
+      print('\r{} records'.format(len(series)), end='')
```
Comment (WilliamCFong) [inline astronet/predict.py]: It's low priority, but I remember this return control character doesn't always refresh the entire line, so the file output is a bit jank.

Maybe the stdout should be moved to logging?

File: astronet/preprocess/generate_input_records.py [hunk 7]
```diff
@@ -14,16 +14,22 @@
...
 import numpy as np
 import pandas as pd
 import tensorflow as tf
+from absl import app, logging
+from typing_extensions import Protocol
 
 from astronet.preprocess import preprocess
 
 
+class LCGetter(Protocol):
```
Comment (WilliamCFong) [inline astronet/preprocess/generate_input_records.py:28]: I keep forgetting about `Protocol`. We should use it in QLP!

File: astronet/direct_tensor/features.py
Comment (havijw) [reply astronet/direct_tensor/features.py:1]: Should include a link to the paper for details on input features.

File: astronet/direct_tensor/predict.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+
+def prediction_features(
+    tce: pd.Series,
+    time: np.ndarray,
+    flux: np.ndarray,
+    aperture_fluxes: dict[str, tuple[np.ndarray, np.ndarray]],
+    breakspaces: list[Optional[float]] = [0.3, 5.0, None],
+):
+    """
+    Assemble all input features for a lightcurve.
```
Comment (havijw) [inline astronet/direct_tensor/predict.py:203]: Should include a link to the paper for details on input features.

File: astronet/direct_tensor/predict.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    Returns
+    -------
+    features: dict[str, float | array[float]]
+        Dict of {feature_name: feature_value} for features that are or depend on views of the
+        lightcurve.
+    fold_num: array[float]
+        Array containing the fold number of each point in the lightcurve. Used to calculate the
+        total number of folds in the lightcurve for the "n_folds" feature.
+    """
+    tag = "" if breakspace is None else f"_{breakspace}"
```
Comment (WilliamCFong) [inline astronet/direct_tensor/predict.py]: Oh, this one is important, at least in my understanding of the astronet model on PDO. The features are expected in the format of `{feature_name}{tag}` or maybe `{feature_name}.{tag}`. I think it's the former.

File: astronet/direct_tensor/predict.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+def find_checkpoints(base_dir: Path, nruns: Optional[int] = None) -> list[Path]:
+    """Find model directories, assuming base directory structure of model training checkpoints."""
+    if nruns is None:
+        nruns = len(list(base_dir.iterdir()))
+    return [next((base_dir / str(i)).iterdir()) for i in range(1, nruns + 1)]
+
+
+def batch_predict(
+    checkpoints_dir: Path,
```
Comment (havijw) [inline astronet/direct_tensor/predict.py:352]: It's probably a good idea to explicitly cast `checkpoints_dir` to `Path` at some point (here or in `find_checkpoints`), since other parts of the codebase use `os.path` so astronet users may want to use string paths rather than `Path` objects.

File: astronet/direct_tensor/predict.py (hunk 9 above)
Comment (WilliamCFong) [reply astronet/direct_tensor/predict.py]: Since it's a common format, I would write a transformer which takes the features and updates the keys using the `tag` (or just pass the tag into the feature generation functions).

File: astronet/direct_tensor/predict.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+            odd_mask,
+            even_mask,
+            period,
+            duration,
+        )
+    )
+
+    double_fold_time, double_fold_flux, _, _ = (
+        preprocess.phase_fold_and_sort_light_curve(
+            det_time, det_flux, transit_mask, period * 2, epoch - period / 2
```
Comment (havijw) [inline astronet/direct_tensor/predict.py:178]: Why do we subtract `period / 2` from `epoch`?

File: astronet/direct_tensor/predict.py (hunk 2 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py]: Use pandas to do a fast check.

File: astronet/predict.py (hunk 4 above)
Comment (havijw) [reply astronet/predict.py:104]: Pull it out into a util function.

File: astronet/predict.py (hunk 6 above)
Comment (havijw) [reply astronet/predict.py]: Logging or `tqdm` since this is really just trying to be a progress bar.

File: astronet/direct_tensor/predict.py (hunk 9 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py]: The tag is correct, but move it closer to where it's used (at the end of the function). Also, tensorflow doesn't like periods in feature names, so do that normalization here instead of letting tensorflow do it.

File: astronet/direct_tensor/predict.py (hunk 11 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py:178]: Add documentation once I figure it out.

File: astronet/direct_tensor/predict.py (hunk 3 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py]: Actually, move `breakspaces` to a global variable.

File: astronet/direct_tensor/predict.py (hunk 11 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py:178]: It's to center the double period view. There are 2 transits in the view, and if we didn't have this subtraction, one of the transits would be centered while the other was split between the very start and very end of the view.

See [this image](https://content.cld.iop.org/journals/1538-3881/165/3/95/revision1/ajacad85f8_hr.jpg) from the 2023 paper on the most recent model.
Diagram showing architecture of the astronet model.

File: astronet/direct_tensor/predict.py (hunk 2 above)
Comment (havijw) [reply astronet/direct_tensor/predict.py]: This is a dictionary, not a dataframe, so pandas can't be faster than iterating.
